_episodes/l2_02_basic_unit_testing.md

```diff
@@ -0,0 +1,293 @@
+- "We learnt about some of the popular testing frameworks for C++."
+- "We learnt about Googletest and how to use it to write our tests."
+- "We learnt about anatomy of a unit test and how to write effective unit tests."
+- "Finally, we saw an example to demonstrate the above concepts."
+---
+
+## 1. Popular frameworks for testing in C++
+
+There are a number of popular frameworks that can be used for testing in C++. A few of them are mentioned below:-
+
+- [GoogleTest](https://google.github.io/googletest/)
+- [Boost.Test](https://www.boost.org/doc/libs/1_82_0/libs/test/doc/html/index.html)
+- [CppUtest](http://cpputest.github.io/)
+- [Catch2](https://github.com/catchorg/Catch2)
+- [UT++](https://boost-ext.github.io/ut/)
+- [doctest](https://github.com/doctest/doctest)
+
+In this course, we will focus mainly on GoogleTest.
+
+
+
+
+## 2. A brief about GoogleTest
+
+Simply defining, GoogleTest is a testing framework developed by Google's testing technology team to develop and write C++ tests.  GoogleTest offers multiple advantages over other frameworks:-
+
+1. **Comprehensive Features**: Google Test provides a rich set of features, including a wide range of assertion macros, test fixtures, parameterized tests, test discovery, test filtering, and powerful mocking capabilities. It offers a complete testing framework that can handle various testing scenarios.
+2. **Integration with Google Test ecosystem**: Google Test integrates well with other tools and libraries in the Google Test ecosystem. For example, it can be seamlessly used with Google Mock for creating mock objects and behaviours. This integration enhances the testing capabilities and flexibility.
+3. **Large and Active Community**: Google Test has a large and active community of developers. This means that there is ample support available in terms of documentation, tutorials, forums, and online resources. 
+4. **Mature and Stable**: It is a mature and stable framework that has been used extensively in industry projects and open-source software. 
+5. **Wide Platform Support**: Google Test supports multiple platforms, including Windows, Linux, macOS, and various compilers. It is compatible with popular development environments and build systems, making it suitable for a wide range of C++ projects.
+6. **Flexible and Extensible**: Google Test provides flexibility in test organization and customization. It allows you to structure your tests using test cases and test suites. You can also define custom test fixtures and customise test execution and reporting. Additionally, Google Test can be extended with custom assertion macros and utilities to suit your specific testing needs.
+
+
+
+## 3. Writing Unit test using GoogleTest
+
+For this chapter, we will consider the same example `fibonacci.cpp` that we used in the previous chapter. Also, for understanding how to use GoogleTest, we will simply be converting all the tests that we wrote manually in the same file using GoogleTest framework.
+
```

Comment:
I think that before diving into the coding there should be a higher level overview of what we are about to do. Something along the lines of:

```suggestion
For this episode, we will consider the same example `fibonacci.cpp` that we used previously. There we identified 3 possible cases when using the `recursive_fibonacci` function for which we wrote some manual tests:

- `n` equal to 0 or 1
- `n` a negative number
- `n` a positive number greater than 1

To demonstrate how to use GoogleTest, we will simply be converting these tests that we wrote manually in the same file using GoogleTest framework. Section 5 below will describe the anatomy of a test, but first lets see how they look like in the practical case.
```